File: backend/src/main/java/com/example/demo/controller/AuthController.java
// backend\src\main\java\com\example\demo\controller\AuthController.java
package com.example.demo.controller;

import com.example.demo.model.User;
import com.example.demo.repository.UserRepository;
import com.example.demo.security.JwtTokenProvider;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.*;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
    
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;
    
    public AuthController(UserRepository userRepository, 
                         PasswordEncoder passwordEncoder,
                         JwtTokenProvider jwtTokenProvider) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtTokenProvider = jwtTokenProvider;
    }
    
    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody Map<String, String> request) {
        try {
            System.out.println("📝 Register request: " + request);
            
            String email = request.get("email");
            String password = request.get("password");
            String name = request.get("name");
            String phone = request.get("phone");
            String city = request.get("city");
            
            // Валидация
            if (email == null || email.trim().isEmpty()) {
                return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", "Email обязателен"
                ));
            }
            
            if (password == null || password.trim().isEmpty()) {
                return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", "Пароль обязателен"
                ));
            }
            
            // Проверка существования пользователя
            if (userRepository.existsByEmail(email)) {
                return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", "Пользователь с таким email уже существует"
                ));
            }
            
            // Создание нового пользователя
            User user = new User();
            user.setName(name != null ? name : email.split("@")[0]);
            user.setEmail(email.toLowerCase());
            user.setPassword(passwordEncoder.encode(password));
            user.setPhone(phone != null ? phone : "+77000000000");
            user.setCity(city != null ? city : "Алматы");
            user.setBalance(5000.0);
            user.setIsActive(true);
            
            // Сохранение пользователя
            User savedUser = userRepository.save(user);
            System.out.println("✅ User saved with ID: " + savedUser.getId());
            
            // Генерация токена
            String token = jwtTokenProvider.generateToken(
                savedUser.getEmail(), 
                savedUser.getRole().name(),
                savedUser.getId()
            );
            
            // Подготовка данных пользователя для ответа
            Map<String, Object> userData = getUserDataMap(savedUser);
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Регистрация выполнена успешно",
                "token", token,
                "user", userData
            ));
            
        } catch (Exception e) {
            System.err.println("❌ Registration error: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.badRequest().body(Map.of(
                "success", false,
                "message", "Ошибка при регистрации: " + e.getMessage()
            ));
        }
    }
    
    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> request) {
        try {
            System.out.println("🔐 Login request: " + request);
            
            String email = request.get("email");
            String password = request.get("password");
            
            // Валидация
            if (email == null || email.trim().isEmpty()) {
                return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", "Email обязателен"
                ));
            }
            
            if (password == null || password.trim().isEmpty()) {
                return ResponseEntity.badRequest().body(Map.of(
                    "success", false,
                    "message", "Пароль обязателен"
                ));
            }
            
            // Поиск пользователя
            Optional<User> userOptional = userRepository.findByEmail(email.toLowerCase());
            if (userOptional.isEmpty()) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Пользователь не найден"
                ));
            }
            
            User user = userOptional.get();
            
            // Проверка пароля
            if (!passwordEncoder.matches(password, user.getPassword())) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Неверный пароль"
                ));
            }
            
            // Проверка активности аккаунта
            if (!user.getIsActive()) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Аккаунт заблокирован"
                ));
            }
            
            // Обновление времени последнего входа
            user.setUpdatedAt(LocalDateTime.now());
            userRepository.save(user);
            
            // Генерация токена
            String token = jwtTokenProvider.generateToken(
                user.getEmail(), 
                user.getRole().name(),
                user.getId()
            );
            
            // Подготовка данных пользователя для ответа
            Map<String, Object> userData = getUserDataMap(user);
            
            System.out.println("✅ Login successful for user: " + user.getEmail());
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Вход выполнен успешно",
                "token", token,
                "user", userData
            ));
            
        } catch (Exception e) {
            System.err.println("❌ Login error: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.badRequest().body(Map.of(
                "success", false,
                "message", "Ошибка при входе: " + e.getMessage()
            ));
        }
    }
    
    @GetMapping("/me")
    public ResponseEntity<?> getCurrentUser(@RequestHeader("Authorization") String authHeader) {
        try {
            System.out.println("👤 /api/auth/me called, authHeader: " + 
                (authHeader != null ? authHeader.substring(0, Math.min(authHeader.length(), 30)) + "..." : "null"));
            
            // Проверка заголовка авторизации
            if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Токен не предоставлен"
                ));
            }
            
            String token = authHeader.substring(7);
            
            // Валидация токена
            if (!jwtTokenProvider.validateToken(token)) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Недействительный токен"
                ));
            }
            
            // Извлечение userId из токена
            Long userId = jwtTokenProvider.getUserIdFromToken(token);
            if (userId == null) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Не удалось получить данные пользователя из токена"
                ));
            }
            
            // Поиск пользователя в БД
            Optional<User> userOptional = userRepository.findById(userId);
            if (userOptional.isEmpty()) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Пользователь не найден"
                ));
            }
            
            User user = userOptional.get();
            
            // Проверка активности аккаунта
            if (!user.getIsActive()) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Аккаунт заблокирован"
                ));
            }
            
            // Подготовка данных пользователя для ответа
            Map<String, Object> userData = getUserDataMap(user);
            
            System.out.println("✅ /api/auth/me successful for user: " + user.getEmail());
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "user", userData
            ));
            
        } catch (Exception e) {
            System.err.println("❌ /api/auth/me error: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.badRequest().body(Map.of(
                "success", false,
                "message", "Ошибка: " + e.getMessage()
            ));
        }
    }
    
    @PostMapping("/logout")
    public ResponseEntity<?> logout(@RequestHeader("Authorization") String authHeader) {
        try {
            // В реальном приложении здесь можно добавить токен в blacklist
            // Для JWT stateless просто возвращаем успех
            return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Выход выполнен успешно"
            ));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(Map.of(
                "success", false,
                "message", "Ошибка при выходе: " + e.getMessage()
            ));
        }
    }
    
    @PostMapping("/refresh")
    public ResponseEntity<?> refreshToken(@RequestHeader("Authorization") String authHeader) {
        try {
            if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Токен не предоставлен"
                ));
            }
            
            String oldToken = authHeader.substring(7);
            
            if (!jwtTokenProvider.validateToken(oldToken)) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Недействительный токен"
                ));
            }
            
            String email = jwtTokenProvider.getEmailFromToken(oldToken);
            Long userId = jwtTokenProvider.getUserIdFromToken(oldToken);
            String role = jwtTokenProvider.extractRole(oldToken);
            
            if (email == null || userId == null) {
                return ResponseEntity.status(401).body(Map.of(
                    "success", false,
                    "message", "Не удалось получить данные из токена"
                ));
            }
            
            // Генерация нового токена
            String newToken = jwtTokenProvider.generateToken(email, role, userId);
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "token", newToken,
                "message", "Токен обновлен"
            ));
            
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(Map.of(
                "success", false,
                "message", "Ошибка при обновлении токена: " + e.getMessage()
            ));
        }
    }
    
    /**
     * Вспомогательный метод для преобразования User в Map
     */
    private Map<String, Object> getUserDataMap(User user) {
        Map<String, Object> userData = new HashMap<>();
        userData.put("id", user.getId());
        userData.put("name", user.getName());
        userData.put("email", user.getEmail());
        userData.put("phone", user.getPhone());
        userData.put("city", user.getCity());
        userData.put("role", user.getRole().name());
        userData.put("balance", user.getBalance());
        userData.put("avatarUrl", user.getAvatarUrl());
        userData.put("address", user.getAddress());
        userData.put("latitude", user.getLatitude());
        userData.put("longitude", user.getLongitude());
        userData.put("companyName", user.getCompanyName());
        userData.put("bin", user.getBin());
        userData.put("isActive", user.getIsActive());
        userData.put("createdAt", user.getCreatedAt());
        userData.put("updatedAt", user.getUpdatedAt());
        
        // Для совместимости с фронтендом
        userData.put("nickname", user.getName());
        userData.put("avatar", user.getAvatarUrl());
        
        return userData;
    }
}

File: nexus/src/auth/auth.controller.js
const { validationResult } = require('express-validator');

class AuthController {
    constructor(authService) {
        this.authService = authService;
    }

    async register(req, res) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    message: 'Ошибка валидации',
                    errors: errors.array()
                });
            }

            const { name, email, password, phone, city, address, confirmPassword } = req.body;

            if (password !== confirmPassword) {
                return res.status(400).json({
                    success: false,
                    message: 'Пароли не совпадают'
                });
            }

            console.log(`📝 Регистрация: ${email}`);

            const result = await this.authService.register({
                name: name.trim(),
                email: email.trim().toLowerCase(),
                password,
                phone: phone || null,
                city: city || null,
                address: address || null
            });

            console.log(`✅ Регистрация успешна: ${email}, ID: ${result.user.id}`);

            res.status(201).json({
                success: true,
                message: 'Регистрация прошла успешно',
                data: result
            });

        } catch (error) {
            console.error('❌ Ошибка регистрации:', error.message);

            if (error.code === 'DUPLICATE_EMAIL') {
                return res.status(409).json({
                    success: false,
                    message: error.message
                });
            }

            res.status(500).json({
                success: false,
                message: 'Ошибка сервера при регистрации'
            });
        }
    }

    async login(req, res) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    message: 'Ошибка валидации',
                    errors: errors.array()
                });
            }

            const { email, password } = req.body;

            console.log(`🔑 Вход: ${email}`);

            const result = await this.authService.login(
                email.trim().toLowerCase(), 
                password
            );

            console.log(`✅ Вход успешен: ${email}, ID: ${result.user.id}`);

            res.json({
                success: true,
                message: 'Вход выполнен успешно',
                data: result
            });

        } catch (error) {
            console.error('❌ Ошибка входа:', error.message);

            if (error.code === 'INVALID_CREDENTIALS') {
                return res.status(401).json({
                    success: false,
                    message: error.message
                });
            }

            res.status(500).json({
                success: false,
                message: 'Ошибка сервера при входе'
            });
        }
    }

    async verify(req, res) {
        try {
            const token = req.headers.authorization?.replace('Bearer ', '');
            
            if (!token) {
                return res.status(401).json({
                    success: false,
                    message: 'Токен не предоставлен'
                });
            }

            const user = await this.authService.getUserFromToken(token);

            res.json({
                success: true,
                data: { user }
            });

        } catch (error) {
            console.error('❌ Ошибка верификации токена:', error.message);

            if (error.code === 'TOKEN_EXPIRED' || error.code === 'INVALID_TOKEN') {
                return res.status(401).json({
                    success: false,
                    message: error.message
                });
            }

            res.status(500).json({
                success: false,
                message: 'Ошибка сервера при верификации токена'
            });
        }
    }

    async getMe(req, res) {
        try {
            if (!req.user) {
                return res.status(401).json({
                    success: false,
                    message: 'Пользователь не авторизован'
                });
            }

            const user = await this.authService.userModel.findById(req.user.id);

            if (!user) {
                return res.status(404).json({
                    success: false,
                    message: 'Пользователь не найден'
                });
            }

            res.json({
                success: true,
                data: { user }
            });

        } catch (error) {
            console.error('❌ Ошибка получения информации о пользователе:', error);
            
            res.status(500).json({
                success: false,
                message: 'Ошибка сервера'
            });
        }
    }

    async changePassword(req, res) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    errors: errors.array()
                });
            }

            const { currentPassword, newPassword } = req.body;
            const userId = req.user?.id;

            if (!userId) {
                return res.status(401).json({
                    success: false,
                    message: 'Пользователь не авторизован'
                });
            }

            await this.authService.changePassword(userId, currentPassword, newPassword);

            res.json({
                success: true,
                message: 'Пароль успешно изменен'
            });

        } catch (error) {
            console.error('❌ Ошибка смены пароля:', error);
            
            if (error.message === 'Неверный текущий пароль') {
                return res.status(400).json({
                    success: false,
                    message: error.message
                });
            }

            res.status(500).json({
                success: false,
                message: 'Ошибка сервера при смене пароля'
            });
        }
    }

    async requestPasswordReset(req, res) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    message: 'Ошибка валидации',
                    errors: errors.array()
                });
            }

            const { email } = req.body;
            
            console.log(`📧 Запрос восстановления пароля для: ${email}`);

            const result = await this.authService.requestPasswordReset(email);

            console.log(`✅ Код отправлен на: ${email}`);

            res.json({
                success: true,
                message: result.message
            });

        } catch (error) {
            console.error('❌ Ошибка запроса восстановления пароля:', error.message);

            if (error.code === 'USER_NOT_FOUND') {
                return res.status(404).json({
                    success: false,
                    message: error.message
                });
            }

            res.status(500).json({
                success: false,
                message: error.message || 'Ошибка при отправке кода'
            });
        }
    }

    async verifyResetCode(req, res) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    message: 'Ошибка валидации',
                    errors: errors.array()
                });
            }

            const { email, code } = req.body;
            
            console.log(`🔐 Проверка кода для: ${email}`);

            const result = await this.authService.verifyResetCode(email, code);

            console.log(`✅ Код подтвержден для: ${email}`);

            res.json({
                success: true,
                message: 'Код подтвержден',
                token: result.token
            });

        } catch (error) {
            console.error('❌ Ошибка проверки кода:', error.message);

            if (error.code === 'INVALID_CODE') {
                return res.status(400).json({
                    success: false,
                    message: error.message
                });
            }

            res.status(500).json({
                success: false,
                message: 'Ошибка при проверке кода'
            });
        }
    }

    async resetPassword(req, res) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    message: 'Ошибка валидации',
                    errors: errors.array()
                });
            }

            const { email, token, newPassword } = req.body;

            if (!newPassword || newPassword.length < 6) {
                return res.status(400).json({
                    success: false,
                    message: 'Пароль должен содержать минимум 6 символов'
                });
            }

            console.log(`🔄 Сброс пароля для: ${email}`);

            const result = await this.authService.resetPassword(email, token, newPassword);

            console.log(`✅ Пароль изменен для: ${email}`);

            res.json({
                success: true,
                message: result.message
            });

        } catch (error) {
            console.error('❌ Ошибка сброса пароля:', error.message);

            if (error.message.includes('истек') || error.message.includes('Неверный токен')) {
                return res.status(400).json({
                    success: false,
                    message: error.message
                });
            }

            res.status(500).json({
                success: false,
                message: 'Ошибка при сбросе пароля'
            });
        }
    }
}

module.exports = AuthController;